reference/src/layout/unions.md

```diff
@@ -6,29 +6,107 @@ not to change until an RFC ratifies them.
 
 [#13]: https://github.com/rust-rfcs/unsafe-code-guidelines/issues/13
 
-The only degree of freedom the compiler has when computing the layout of a union
-like
+### Layout of individual union fields
 
-```rust,ignore
-union U { f1: T1, f2: T2 }
+A union consists of several variants, one for each field. All variants have the
+same size and start at the same memory address, such that in memory the variants
+overlap. This can be visualized as follows:
+
+```text
+[ <--> [field0_ty] <----> ]
+[ <----> [field1_ty] <--> ]
+[ <---> [field2_ty] <---> ]
 ```
+**Figure 1** (union-field layout): Each row in the picture shows the layout of
+the union for each of its variants. The `<-...->` and `[ ... ]` denote the
+differently-sized gaps and fields, respectively.
+
+The individual fields (`[field{i}_ty_]`) are blocks of fixed size determined by
+the field's [layout]. The only degrees of freedom the compiler has when
+computing the layout of a union are the size of the union, which can be larger
+than the size of its largest field, and the offset of each union field within
+its variant. How these are picked depends on certain constraints like, for
+example, the alignment requirements of the fields, the `#[repr]` attribute of
+the `union`, etc.
```

Comment:
For structs, my PR says something like

> The layout of these fields
themselves is already entirely determined by their types, and since we intend to
allow creating references to fields (`&s.f1`), structs do not have any
wiggle-room there.

Might be worth also adding a sentence here justifying why that is the only freedom the compiler has.

Reply:
I've squeezed that in a bit.